core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -3801,6 +3803,107 @@ private ExpressionNode rewriteAggregate(ExpressionNode agg, QueryModel model) {
+     SELECT timestamp from (SELECT timestamp from table_name) limit 1
+     This disables invoking group by workers
+     */
+    private void optimiseAggregateFunctions(QueryModel parent){
+
+        //base condition to stop recursion
+        if(parent== null || parent.getNestedModel() == null)
+          return;
+
+        ObjList<QueryColumn> queryColumns = parent.getBottomUpColumns();
+        QueryModel nestedModel = parent.getNestedModel();
+        CharSequence designatedTimestampColumn = null;
+
+        /**if FIRST/LAST/min/max(column) does not contain designated-timestamp column
+         * OR
+         * another column present in select clause
+         * then don't apply optimisations to base model
+         */
+        if (nestedModel.getTimestamp() == null || queryColumns.size() > 1) {
+            optimiseAggregateFunctions(nestedModel);
+            return;
+        }
+
+        //get designated timestamp column name
+        designatedTimestampColumn =  nestedModel.getTimestamp().token;
+        QueryColumn column = queryColumns.get(0);
+        ExpressionNode ast = column.getAst();
+        CharSequence token= null;
+        CharSequence rhs= null;
+        if(ast != null) {
+            token = ast.token;
+             rhs = ast.rhs == null ? null: ast.rhs.token;
+        }
+
+        /**
+         Type 1 Optimisations: Will be optimised by adding order by clause and changing model type
+         Type 2 Optimisations: Will be done only by changing model type to prevent invoking group by workers
+         */
+        int optimisationType = 0;
+        if(AGGREGATE_FUNCTIONS_OPTIMISED_BY_BACKWARD_SCAN.contains(token.toString()))
+            optimisationType = 1;
+        else if(AGGREGATE_FUNCTIONS_OPTIMISED_BY_FORWARD_SCAN.contains(token.toString()))
+            optimisationType = 2;
+
+        /**
+         Core logic for issue #4231 will be applicable under two conditions
+         Condition1: QueryColumn contains FIRST/LAST/max/min function keyword
+         Condition2: QueryColumn for LAST/max keyword should be same as designatedTimestamp column
+         Core logic for changing query plan
+         - Add limit 1 via expressionNode to parent model
+         - Change model to select-choose from select-group-by
+         - Column alias by default is LAST/max if alias is not provided, replace with original column name
+         - Add order by clause to nested model
+         Call this recursively for nested models
+         */
+        if (rhs!=null && (optimisationType == 1 || optimisationType == 2)
+            && designatedTimestampColumn.toString().equals(rhs.toString())) {
+            switch (optimisationType) {
+                case 1:
+                    ExpressionNode newTimestampNode =  expressionNodePool.next();
+                    newTimestampNode.token = designatedTimestampColumn;
+                    nestedModel.addOrderBy(newTimestampNode, QueryModel.ORDER_DIRECTION_DESCENDING);
```

Comment:
It's better to create a new model rather than mutate the old one.

Reply:
@puzpuzpuz is it expected to create a new model using QueryModelFactory and copy columns from old nestedModel? 
I can see old being mutated in many places in optimiser

Reply:
Creating a new model is as simple as `queryModelPool.next()`. Mutating an existing model should be done with extra care, so we only do it when we change a part of the model.